add tests for ddp with activation check points

zhaojuanmao opened the pull request:
Summary: add two success cases and two failure cases for ddp with activation check points when grad_as_bucket_view = true and false

Test Plan: unit tests

Differential Revision: D26679895

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+                                                    process_group=process_group,
+                                                    find_unused_parameters=find_unused_parameters)
+        input_tensor = torch.rand((64, 2000)).cuda()
+        input_tensor.requires_grad = True
```
nit: I think you can specify requires_grad=True in constructor of tensor itself.

mcarilli replied:
Cuda too, and we need to make sure not a record a copy to the GPU in the autograd history. Use the one-liner
```python
input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
```

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+                                                 use_bucket_view=use_bucket_view)
+            norm = 0.0
+            for p in model.parameters():
+                assert p.grad is not None
```
Use `self.assertTrue` here?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+            for p in model.parameters():
+                assert p.grad is not None
+                norm += p.grad.norm().item()
+            assert numpy.allclose(norm, 78053.52966308594), norm
```
What does `assert numpy.allclose(norm, 78053.52966308594), norm` check for? Also, could we analytically compute the norm `78053` instead of hardcoding it?

zhaojuanmao replied:
did not know how to calculate the norm..., do you have reference? the test should be good as we set the same seed "torch.manual_seed(31415)" to make sure results are reproducible.

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+            output_tensor.sum().backward()
+            norm = 0.0
+            for p in model.parameters():
+                assert p.grad is not None
```
same as above

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+            for p in model.parameters():
+                assert p.grad is not None
+                norm += p.grad.norm().item()
+            assert numpy.allclose(norm, 57004.33825683594), norm
```
same as above

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -1190,20 +1190,12 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
-                    "to point to bucket_view_in0, and hence hit this "
-                    "warning. If you saw this message, please double-check if "
-                    "the above situation is expected for your application.");
+                    "expected DDP bucket view with gradient_as_bucket_view=True. "
```
not from this PR, but above line should end with " _where_ gradient is not the ..."

rohan-varma reviewed (approved):
Wonder if it's worth it to also have a more general deduplication check (unittest)? This might help us eliminate duplication bugs if the possibility comes up in another bug. 

Something like, have a model with duplicate parameters, create `Reducer`, and expose methods to access its parameters and ensure that they are deduped.

mcarilli commented on test/distributed/test_c10d.py:
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
+                                                        device_ids=[self.rank],
+                                                        process_group=process_group)
+            input_tensor = torch.rand((64, 2000)).cuda()
+            input_tensor.requires_grad = True
```
```python
input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
```

zhaojuanmao commented:
@rohan-varma regarding deduplication check (or unitest), I'm planning to revert the changes in  https://github.com/pytorch/pytorch/pull/46755 as right now we deduplicated params. along with it, I will add an assert if it is found that duplicated parameters sharing the same grad accumulator in the gradAccToVariableMap_ . ==> that should be sufficient for deduplication check

mcarilli commented:
Do not merge yet, https://github.com/pytorch/pytorch/pull/51929/ got temporarily reverted (see https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662). Give us a while to track down source of asan failures.

mcarilli reviewed (commented):
"requesting changes" as a reminder not to land until we investigate why https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662 got temporarily reverted (see the last comment on that PR, for some reason linking directly to it doesn't work).

mcarilli reviewed (changes requested):
see https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662

zhaojuanmao committed:
add tests for ddp with activation check points (#52894)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/52894

add two success cases and two failure cases for ddp with activation check points when grad_as_bucket_view = true and false

Test Plan: unit tests

Reviewed By: rohan-varma

Differential Revision: D26679895

fbshipit-source-id: 85d85c401e99991447849f8079bc7e1aaebfd5a3
